src/common/src/array/stream_chunk_builder.rs

```diff
@@ -0,0 +1,146 @@
+            column_builders,
+            data_types,
+            capacity: chunk_size,
+            size: 0,
+        }
+    }
+
+    /// Increase chunk size
+    ///
+    /// A [`StreamChunk`] will be returned when `size == capacity`
+    #[must_use]
+    fn inc_size(&mut self) -> Option<StreamChunk> {
+        self.size += 1;
+
+        // Take a chunk when capacity is exceeded. Splitting `UpdateDelete` and `UpdateInsert`
+        // should be avoided, so when the last one is `UpdateDelete`, we delay the chunk until
+        // `UpdateInsert` comes. This means the output chunk size may exceed the given `chunk_size`,
+        // and theoretically at most `chunk_size + 1` if inputs are consistent.
+        if self.size >= self.capacity && self.ops[self.ops.len() - 1] != Op::UpdateDelete {
+            self.take()
+        } else {
+            None
+        }
+    }
+
+    /// Append an iterator of output index and datum to the builder, return a chunk if the builder
+    /// is full.
+    ///
+    /// Note: the caller must ensure that each column occurs exactly once in `iter`.
+    #[must_use]
+    pub fn append_iter<'a>(
+        &mut self,
+        op: Op,
+        iter: impl IntoIterator<Item = (usize, DatumRef<'a>)>,
+    ) -> Option<StreamChunk> {
+        self.ops.push(op);
+        for (i, datum) in iter {
+            self.column_builders[i].append(datum);
+        }
+        self.inc_size()
+    }
+
+    /// Append a row to the builder, return a chunk if the builder is full.
+    #[must_use]
+    pub fn append_row(&mut self, op: Op, row: impl Row) -> Option<StreamChunk> {
+        self.append_iter(op, row.iter().enumerate())
+    }
+
+    /// Append a record to the builder, return a chunk if the builder is full.
+    #[must_use]
+    pub fn append_record(&mut self, record: Record<impl Row>) -> Option<StreamChunk> {
+        match record {
+            Record::Insert { new_row } => self.append_row(Op::Insert, new_row),
+            Record::Delete { old_row } => self.append_row(Op::Delete, old_row),
+            Record::Update { old_row, new_row } => {
+                let none = self.append_row(Op::UpdateDelete, old_row);
+                debug_assert!(none.is_none());
+                self.append_row(Op::UpdateInsert, new_row)
+            }
+        }
+    }
+
+    #[must_use]
+    pub fn take(&mut self) -> Option<StreamChunk> {
+        if self.size == 0 {
+            return None;
+        }
+
+        self.size = 0;
+        let new_columns = self
+            .column_builders
+            .iter_mut()
+            .zip_eq_fast(&self.data_types)
+            .map(|(builder, datatype)| {
+                std::mem::replace(builder, datatype.create_array_builder(self.capacity)).finish()
+            })
+            .map(Into::into)
+            .collect::<Vec<_>>();
+
+        Some(StreamChunk::new(
+            std::mem::replace(&mut self.ops, Vec::with_capacity(self.capacity)),
+            new_columns,
+        ))
+    }
+}
```

Comment:
Are there any diffs to review for this file? Seems like it's entirely migrated from `comm/builder.rs`

Reply:
Yes. There're no changes.